Added Backdrop Node for Meshroom graph

waaake opened the pull request:
## Feature Request
* Added Backdrop Node
* Notion of Region inside a Graph

## Minor Fix
* Added Key Press of Backspace to allow deletion in the graph.

waaake committed:
[desc] Node Internal Attributes: Added Attribute factory

Attribute factory serves as a common place to initialize and group attributes and define an accessor to be able to retrieve the attributes

waaake committed:
[core] Backdrop Node: Added Notion of Backdrop as a Node

Backdrop Node serves as a Graphical Node allowing only internal attributes to be listed and can be serialized

waaake committed:
[desc] Node Attribute Traits: Added Resizable Trait and attribute defines

waaake committed:
[desc] Added Notion of Builtin Backdrop

Added Backdrop Node descriptor which is a derived Incomputable Node with Resizable Attribute Traits

waaake committed:
[ui] Geom2d: Added a way to detect full intersection between two rects

waaake committed:
[ui][qml] Added Backdrop Node component in QML

waaake committed:
[ui] Graph: Added Functionality to modify node position and dimensions

waaake committed:
[ui] GraphEditor: Added Loader for Node Delegate

The Loader allows loading Backdrop Component for backdrop nodes and Standard Node component for other nodes

waaake committed:
[core] Builtins: Added registration mechanism for Builtin plugins

waaake committed:
[core] Node: Added Constructor Preference to allow defining the Node constructor type

waaake committed:
[core] Graph: Updated Node Construction to rely on using preferred constructor for Node Types

waaake committed:
[ui][qml] GraphEditor: Added helper for nodeRepeater to get the item at a given index

With the change in nodeRepeater to hold the Loader instead of active delegate, fetching the item at a given index involves getting the loader at index and then it's item. This function allows easy access to the item directly.

waaake committed:
[ui] GraphEditor: Updated bounding box to look at the actual node item rather than the loader delegate

waaake committed:
[ui] DelegateSelectionBox: Updated selection to refer to the actual item rather than the loader delegate

waaake committed:
[ui] GraphEditor: Grouped the visual resize of a Node and reposition

When dragging a backdrop node from the left side the node has to be moved and resized at the same time and the same needs to be done when an undo is performed, in order to do that with a single undo, a grouping is necessary for the two operations

waaake committed:
[ui] Backdrop: Updated the signal emission from QML to be invoked only when the drag of the backdrop is complete

waaake committed:
[ui] GraphEditor: Backdrop Resizing updates

waaake committed:
[ui] GraphEditor: Added flag for holding the animation state for the UI Graph

The disableAnimation flag would be used in the UI to control when an Animation is to occur

waaake committed:
[ui] Commands: Added Grouped UI Modification Command

The Grouped UI Modfications can be used to disabled overall animations for the UI Graph when commands get executed during both undo and redo operation.

waaake committed:
[ui] Graph: Using Grouped UI Graph Modification to disable animation for qml elements

Used graph's disabledAnimation flag to halt the animations when a backdrop node is resized during an undo.

fabiencastan commented on meshroom/core/desc/node.py:
```diff
@@ -53,6 +52,72 @@ class Node(object):
+    ]
+
+    @classmethod
+    def getInternalParameters(cls, traits: Traits) -> List[Union[StringParam, ColorParam, IntParam, FloatParam]]:
```
I don't see a reason to hardcode a list of specific param types. ` -> List[Union[Attribute]]`

waaake replied:
I was hesitant to import Attribute just for this purpose, but it's updated now.

fabiencastan commented:
By default, when we move the backdrop, the elements within it are also moved. If we use Ctrl+Click, we only move the background, but when we stop dragging the backdrop, the elements in it are still moved!

![pcoip_client_cli_THKCgNzfmI](https://github.com/user-attachments/assets/1707de93-6ed4-4923-af4f-a8b29c391ec7)

fabiencastan commented:
The brightness of the node's default color should remain low for the links to be visible.

fabiencastan commented:
Minimal size for the backdrop is too large:
![image](https://github.com/user-attachments/assets/8112feea-cf17-4556-91d0-b93878f7a9d7)

fabiencastan commented:
Copy-paste does not work:

https://github.com/user-attachments/assets/3956c5b1-550b-45b5-8ce3-3b1407c0ded8

fabiencastan commented on meshroom/core/graph.py:
```diff
@@ -768,7 +768,11 @@ def addNewNode(self, nodeType, name=None, position=None, **kwargs):
             name = self._createUniqueNodeName(name)
 
-        n = self.addNode(Node(nodeType, position=position, **kwargs), uniqueName=name)
+        # Get the Node Constructor which should be initialized for the given node type
+        # Node or Backdrop...
+        NodeType = getPreferredNodeConstructor(nodeType)
+
+        n = self.addNode(NodeType(nodeType, position=position, **kwargs), uniqueName=name)
```
Directly use a function to create the node, instead of a function to return the node type.
```suggestion
        n = self.addNode(createNode(nodeType, position=position, **kwargs), uniqueName=name)
```

waaake replied:
This has been updated.

fabiencastan commented on meshroom/core/desc/builtins.py:
```diff
@@ -0,0 +1,12 @@
+""" Defines the Built in Plugins.
```
I would expect the embedded nodes to be in a folder loaded automatically by Meshroom, in the same way than we are doing for external nodes:
`meshroom/nodes/builtins/Backdrop.py`

waaake replied:
Discussing with @yann-lty initially to have a way to register builtin plugins.
So have added this accordingly.
We can discuss and simplify this if needed 🙂

fabiencastan commented on meshroom/core/__init__.py:
```diff
@@ -357,3 +361,8 @@ def initPipelines():
         else:
             logging.error("Pipeline templates folder '{}' does not exist.".format(f))
+
+def initBuiltinNodePlugins():
+    """ Registers the Builtin plugins for Meshroom.
+    """
+    registerNodeType(builtins.Backdrop)
```
We can remove that if we move it to the standard folder.

fabiencastan commented on meshroom/ui/qml/GraphEditor/Backdrop.qml:
```diff
@@ -0,0 +1,376 @@
+
+        cursorShape: drag.active ? Qt.ClosedHandCursor : Qt.ArrowCursor
+
+        /// Backdrop Resize Controls ???
```
Why "???"

fabiencastan commented on meshroom/ui/qml/GraphEditor/Backdrop.qml:
```diff
@@ -0,0 +1,376 @@
+        }
+
+        ///
+        /// Resize Bottom
```
Probably wants to be able to resize from the top too.

waaake replied:
Resize Top has been added.

fabiencastan commented on meshroom/ui/qml/GraphEditor/Backdrop.qml:
```diff
@@ -0,0 +1,376 @@
+                        root.height = root.height + mouseY;
+
+                        // Ensure a minimum height
+                        if (root.height < 300) {
+                            root.height = 300;
+                        }
```
Probably want to declare this constant somewhere else.
This value is too big.

waaake replied:
set as const value of 200, which seems more reasonable.

fabiencastan commented on meshroom/ui/qml/GraphEditor/Backdrop.qml:
```diff
@@ -0,0 +1,376 @@
+        ///
+        /// Resize Bottom
+        ///
+        Rectangle {
```
It would be good to display a line on hover as we do between top widgets.

waaake commented:
> Minimal size for the backdrop is too large: ![image](https://private-user-images.githubusercontent.com/153585/402256181-8112feea-cf17-4556-91d0-b93878f7a9d7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY3NDQ2ODUsIm5iZiI6MTczNjc0NDM4NSwicGF0aCI6Ii8xNTM1ODUvNDAyMjU2MTgxLTgxMTJmZWVhLWNmMTctNDU1Ni05MWQwLWI5Mzg3OGY3YTlkNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDExM1QwNDU5NDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02NzEwZDVlOWUyMTc4M2JmYjkwOGMwZTU4YmZiODRmNDAzOGVlNGJiMDBmZDM3ZjIyOTJmMGZjZWFjMDYxZTYxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.VHuRoyv2dpGGuG-c7nYzDFlHJk7xt0PZue-ApawJQWY)

This is updated to be a const value set to 200 w and 200 h by default

waaake commented:
> Copy-paste does not work:
> 
>  pcoip_client_cli_RbRMhBQYbj.mp4
Aah, I had missed to check this, my bad !!
This is updated to now copy-paste the backdrop along with the contents of it.

waaake commented:
> By default, when we move the backdrop, the elements within it are also moved. If we use Ctrl+Click, we only move the background, but when we stop dragging the backdrop, the elements in it are still moved!
> 
> ![pcoip_client_cli_THKCgNzfmI](https://private-user-images.githubusercontent.com/153585/402254927-1707de93-6ed4-4923-af4f-a8b29c391ec7.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY3NDQ2ODUsIm5iZiI6MTczNjc0NDM4NSwicGF0aCI6Ii8xNTM1ODUvNDAyMjU0OTI3LTE3MDdkZTkzLTZlZDQtNDkyMy1hZjRmLWE4YjI5YzM5MWVjNy5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDExM1QwNDU5NDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jZDE3ZGFhMjNiYjkzMmRjZjc5OWNkMjkzYzg1ZmJlYTM5MzQ2OTkxNGVlYTNjZWQ0ZmVlNmNhN2NkNjhkOTZmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ucsDbfqoanW7GMf9UJPmA7i-53AXL2M53G5xaErGYIE) [ ![pcoip_client_cli_THKCgNzfmI](https://private-user-images.githubusercontent.com/153585/402254927-1707de93-6ed4-4923-af4f-a8b29c391ec7.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY3NDQ2ODUsIm5iZiI6MTczNjc0NDM4NSwicGF0aCI6Ii8xNTM1ODUvNDAyMjU0OTI3LTE3MDdkZTkzLTZlZDQtNDkyMy1hZjRmLWE4YjI5YzM5MWVjNy5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDExM1QwNDU5NDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jZDE3ZGFhMjNiYjkzMmRjZjc5OWNkMjkzYzg1ZmJlYTM5MzQ2OTkxNGVlYTNjZWQ0ZmVlNmNhN2NkNjhkOTZmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ucsDbfqoanW7GMf9UJPmA7i-53AXL2M53G5xaErGYIE) ](https://private-user-images.githubusercontent.com/153585/402254927-1707de93-6ed4-4923-af4f-a8b29c391ec7.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY3NDQ2ODUsIm5iZiI6MTczNjc0NDM4NSwicGF0aCI6Ii8xNTM1ODUvNDAyMjU0OTI3LTE3MDdkZTkzLTZlZDQtNDkyMy1hZjRmLWE4YjI5YzM5MWVjNy5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDExM1QwNDU5NDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jZDE3ZGFhMjNiYjkzMmRjZjc5OWNkMjkzYzg1ZmJlYTM5MzQ2OTkxNGVlYTNjZWQ0ZmVlNmNhN2NkNjhkOTZmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ucsDbfqoanW7GMf9UJPmA7i-53AXL2M53G5xaErGYIE) [ ](https://private-user-images.githubusercontent.com/153585/402254927-1707de93-6ed4-4923-af4f-a8b29c391ec7.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY3NDQ2ODUsIm5iZiI6MTczNjc0NDM4NSwicGF0aCI6Ii8xNTM1ODUvNDAyMjU0OTI3LTE3MDdkZTkzLTZlZDQtNDkyMy1hZjRmLWE4YjI5YzM5MWVjNy5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDExM1QwNDU5NDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jZDE3ZGFhMjNiYjkzMmRjZjc5OWNkMjkzYzg1ZmJlYTM5MzQ2OTkxNGVlYTNjZWQ0ZmVlNmNhN2NkNjhkOTZmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ucsDbfqoanW7GMf9UJPmA7i-53AXL2M53G5xaErGYIE)

Currently the standard Nodes also have this behaviour, can be fixed in a separate one if you say or in this ?

waaake committed:
[desc] Node: Updated return hint -> List[Attribute] on getInternalParams

waaake committed:
[core] Node: Add method to create a node instance in memory

[core] Graph: Updated addNode and pasteNode functions to refer to createNode method from core Node

waaake committed:
[ui] Backdrop: Backdrop can be resized from the top as well

waaake committed:
[ui] GraphEditor: Backdrop Node Animations considers reszing on Y and halts animation while resizing operations

waaake committed:
[ui] Node: Added property denoting if the node delegate is a backdrop

waaake committed:
[ui] Backdrop: Added getters for children indices from a backdrop

updateChildren gets invoked when the node is pressed resulting in an update call before the selection of the children nodes could happen

waaake committed:
[ui] GraphEditor: Selection of children when a backdrop is selected

Moving backdrop now directly moves the selected children explicitly

waaake committed:
[ui] DelegateSelectionBox: Fetch children indices when box selecting backdrops

This allows selecting children explicitly during this call rather than implicitly invoking select children which was inducing a binding loop on selection property

waaake committed:
[ui] Graph: Remove unused moveNodesBy function

With Selection approach on children of backdrop, selected nodes can directly be moved without having the need for a specific function
